rust/frontend/src/function.rs

```diff
@@ -0,0 +1,476 @@
+    let list = list
+        .iter()
+        .map(|&p| unsafe { CStr::from_ptr(p) })
+        .map(|cs| cs.to_bytes())
+        .map(|bs| str::from_utf8(bs).unwrap())
+        .collect();
+    Mutex::new(list)
+}
+
+/// Returns a registered TVM function by name.
+pub fn get_global_func(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+    let name = CString::new(name).unwrap();
+    let mut handle = ptr::null_mut() as ts::TVMFunctionHandle;
+    check_call!(ts::TVMFuncGetGlobal(
+        name.as_ptr() as *const c_char,
+        &mut handle as *mut _
+    ));
+    if !(handle.is_null()) {
+        mem::forget(name);
+        return Some(Function::new(handle, is_global, false));
+    } else {
+        if allow_missing {
+            return None;
+        } else {
+            panic!("Cannot find global function {:?}", name);
+        }
+    }
+}
+
+/// Wrapper around TVM function handle.
+#[derive(Debug, Hash)]
+pub struct Function {
+    handle: ts::TVMFunctionHandle,
+    is_global: bool,
+    is_released: bool,
+}
+
+impl Function {
+    pub(crate) fn new(handle: ts::TVMFunctionHandle, is_global: bool, is_released: bool) -> Self {
+        Function {
+            handle: handle,
+            is_global: is_global,
+            is_released: is_released,
+        }
+    }
+
+    /// For a given function, it returns a function by name.
+    pub fn get_function(name: &str, is_global: bool, allow_missing: bool) -> Option<Function> {
+        let name = CString::new(name).unwrap();
+        GLOBAL_FUNCTION_NAMES
+            .lock()
+            .unwrap()
+            .iter()
+            .find(|&&s| s == name.to_str().expect("Invalid UTF-8 name"))
+            .map(|nm| get_global_func(nm, is_global, allow_missing).unwrap())
+    }
+
+    /// Returns the underlying TVM function handle.
+    pub fn as_handle(&self) -> ts::TVMFunctionHandle {
+        self.handle
+    }
+
+    pub fn is_global(&self) -> bool {
+        self.is_global
+    }
+
+    pub fn is_released(&self) -> bool {
+        self.is_released
+    }
+}
+
+impl Clone for Function {
+    fn clone(&self) -> Function {
+        if !self.is_released {
+            Self {
+                handle: self.handle,
+                is_global: true,
+                is_released: false,
+            }
+        } else {
+            panic!("Released function cannot be cloned!");
+        }
+    }
+}
+
+impl Drop for Function {
+    fn drop(&mut self) {
+        if !self.is_released & !self.is_global {
+            check_call!(ts::TVMFuncFree(self.handle));
+            self.is_released = true;
+        }
+    }
+}
+
+/// Function builder in order to create and call functions.
+/// *Note:* TVM only accepts at most one return value.
+#[derive(Debug, Clone, Default)]
+pub struct Builder<'a> {
+    pub func: Option<Function>,
+    pub arg_buf: Option<Box<[TVMArgValue<'a>]>>,
+    pub ret_buf: Option<Box<[TVMRetValue<'a>]>>,
+}
+
+impl<'a> Builder<'a> {
```

Comment:
can you try to add a macro akin to the runtime's `call_packed!` that takes a variadic list of args and automatically creates the `Builder` invocation? The frequency with which the frontend creates and consumes `Builder` structs creates enough boilerplate that even a proc-macro would be well motivated (it can almost certainly be done using declarative macros, though).

Reply:
Per your previous suggestion, isn't `tvm_call!` a better macro name along with `register_global_func` (with similar python decoration)? declarative macros is definitely a good option. I'll make that for further enhancements after this is merged. So far, some follow-up enhancements such as 2018 editions have come up. I'll create separate follow-up issues then.

Reply:
the name is up to you. the idea is just that you want to make the builder using a macro. I can help out with macros if you'd like, but I'll wait until you push your changes. Would you mind pushing more frequently? It's not in either of our best interests to be reviewing 4k line diffs.

Reply:
I've been pushing into my [repo](https://github.com/ehsanmok/tvm-rust), to make the changes easier to handle and not to consume CI resources until I'm almost done. Some minor refactoring and `TVMRetValue` are left now.